src/string_pool.cpp

```diff
@@ -14,99 +14,87 @@
 
+
 struct pool {
-    char* _p;
+    char* _p{nullptr};
     std::size_t _n{0};
     std::vector<std::unique_ptr<char[]>> _ponds;
+    std::size_t _total_alloc = 0;
 
-    auto empool(std::string_view incoming) {
+    const char* empool(std::string_view incoming) {
         constexpr auto default_min_k = 16 * 1024 * 1024; // 16MB
-        const auto sz = incoming.size();
-        const auto tsz = sz + 1;
+        const uint32_t sz = (uint32_t)incoming.size();
+        const uint32_t tsz = sz + sizeof(uint32_t) + sizeof(size_t) + 1;
+        
         if (_n < tsz) {
             _n = std::max<std::size_t>(default_min_k, tsz);
+            _total_alloc += _n;
             _ponds.push_back(std::make_unique<char[]>(_n));
             _p = _ponds.back().get();
         }
-        std::memcpy(_p, incoming.data(), sz);
-        std::string_view result(_p, sz);
+        size_t h = std::hash<std::string_view>{}(incoming);
+        std::memcpy(_p, &sz, sizeof(uint32_t));
+        std::memcpy(_p + sizeof(uint32_t), &h, sizeof(size_t));
+        std::memcpy(_p + sizeof(uint32_t) + sizeof(size_t), incoming.data(), sz);
+        *(_p + tsz - 1) = 0;
+        
+        const char* result = _p + sizeof(uint32_t) + sizeof(size_t);
         _n -= tsz;
-        _p += sz;
-        *_p++ = 0;
+        _p += tsz;
         return result;
     }
 };
+ 
 
 /**************************************************************************************************/
 
-auto& pool() {
-#if ORC_FEATURE(LEAKY_MEMORY)
-    thread_local struct pool& instance = *(new struct pool());
-#else
-    thread_local struct pool instance;
-#endif // ORC_FEATURE(LEAKY_MEMORY)
-
-    return instance;
-}
-
-/**************************************************************************************************/
-
-struct pool_string_hash {
-    auto operator()(const pool_string& x) const {
-        return x.hash();
-    }
+// The key is the hash; map one to the other.
+struct pool_key_to_hash {
+    auto operator()(size_t key) const { return key;}
 };
 
-struct pool_string_equal_to {
-    auto operator()(const pool_string& x, const pool_string& y) const {
-        return x.hash() == y.hash();
-    }
-};
-
-auto& set() {
-    using set_type = std::unordered_set<pool_string, pool_string_hash, pool_string_equal_to>;
-
-#if ORC_FEATURE(LEAKY_MEMORY)
-    thread_local set_type& instance = *(new set_type());
-#else
-    thread_local set_type instance;
-#endif
-
-    return instance;
-}
-
 /**************************************************************************************************/
 
 } // namespace
 
 /**************************************************************************************************/
 
 pool_string empool(std::string_view src) {
-    auto& s = set();
-    pool_string needle(src);
-    auto found = s.find(needle);
-
-    if (found != s.end()) {
-        return *found;
+    if (src.empty())
+        return pool_string(nullptr);
```

Comment:
good catch

Reply:
turns out to not quite be correct - a string_view can be empty (and a default is returned) but a pool_string is defined so that _data=nullptr is empty, and anything else must have content.

Reply:
Can you add that detail to the comments in the string pool header, then, or here? I could see someone trying to optimize this line later and breaking intent.

Reply:
Added comment